compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIENoAliasFunctionArguments.cpp

```diff
@@ -0,0 +1,166 @@
+    functionsAndCallers.push_back({funcOp, callers});
+  });
+
+  // Add the callers to the mapping `functionsAndCallers`
+  rootOp->walk([&](func::CallOp callOp) {
+    StringRef callee = callOp.getCallee();
+    auto it = funcOpIndex.find(callee);
+    if (it != funcOpIndex.end()) {
+      functionsAndCallers[it->second].second.push_back(callOp);
+    }
+  });
+  return functionsAndCallers;
+}
+
+/// Traverse backwards through the definition chain of first operands
+/// starting from `initial`, until either a memref.alloc or a amdaie.buffer
+/// operation is found. If neither is found, return a failure.
+FailureOr<Operation *> getDefiningAllocation(Operation *initial) {
+  Operation *current = initial;
+  while (current) {
+    if (isa<AMDAIE::BufferOp>(current) || isa<memref::AllocOp>(current)) {
+      return current;
+    }
+    if (current->getNumOperands() != 1) {
+      InFlightDiagnostic message =
+          initial->emitOpError()
+          << "could not be traced back to an allocation operation, "
+          << "an operation with " << current->getNumOperands()
+          << " operands was encountered while traversing defining ops.";
+      return message;
+    }
+    current = current->getOperand(0).getDefiningOp();
+  }
+  return initial->emitOpError()
+         << "could not be traced back to an allocation operation.";
+};
+
+/// Return a vector containing for every operand of `callOp`, a bool that is
+/// true if the operand is an alias of a memref that does not alias with any
+/// other operand.
+FailureOr<SmallVector<bool>> nonAliasingMemrefArguments(func::CallOp callOp) {
+  // Find the allocations that define the memref operands of the call op.
+  // This vector contains, for each memref operand, a pair containing
+  // the allocation that defines it, and the index of the operand.
+  SmallVector<std::pair<Operation *, uint32_t>> memrefs;
+  for (auto enOperand : llvm::enumerate(callOp.getOperands())) {
+    Value operand = enOperand.value();
+    if (!isa<MemRefType>(operand.getType())) continue;
+    if (operand.getDefiningOp() == nullptr) {
+      return callOp->emitOpError(
+          "has an operand with no defining op, failed to find allocation");
+    }
+    FailureOr<Operation *> maybeAllocation =
+        getDefiningAllocation(operand.getDefiningOp());
+    if (failed(maybeAllocation)) return failure();
+    Operation *allocation = maybeAllocation.value();
+    memrefs.push_back({allocation, enOperand.index()});
+  }
+
+  SmallVector<bool> nonAliasingMemref(callOp.getNumOperands(), false);
+  for (auto memref : memrefs) {
+    bool isAliasing = false;
+    for (auto otherMemref : memrefs) {
+      if (memref.second != otherMemref.second &&
+          memref.first == otherMemref.first) {
+        isAliasing = true;
+      }
+    }
+    uint32_t operandIndex = memref.second;
+    nonAliasingMemref[operandIndex] = !isAliasing;
+  }
+  return nonAliasingMemref;
+}
+
+class AMDAIENoAliasFunctionArgumentsPass
```

Comment:
```suggestion
class AMDAIEAddNoAliasToFunctionArgumentsPass
```
I would prefer starting with a verb as I think it makes the purpose more clear and for consistency with other passes (not all but most of our passes use this format).